go-client/client/client.go

```diff
@@ -0,0 +1,262 @@
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+// Returned as an error when trying to perform a network operation on a client that's been closed.
+var ErrClosedClient = errors.New("client is closed")
+
+type fieldId struct {
+	appId     string
+	fieldName string
+}
+
+// Maintains a connection to a Deephaven server.
+// It can be used to run scripts, create new tables, execute queries, etc.
+// Check the various methods of Client to learn more.
+type Client struct {
+	grpcChannel *grpc.ClientConn
+
+	sessionStub
+	consoleStub
+	flightStub
+	tableStub
+	appStub
+	inputTableStub
+
+	nextTicket int32
+
+	tablesLock sync.Mutex
+	tables     map[fieldId]*TableHandle
+}
+
+// Starts a connection to a Deephaven server.
+// scriptLanguage can be either "python" or "groovy", and must match the language used on the server. Python is the default.
+//
+// The client should be closed using Close() after it is done being used.
+//
+// Note that the provided context is saved and used to send keepalive messages.
+func NewClient(ctx context.Context, host string, port string, scriptLanguage string) (*Client, error) {
+	grpcChannel, err := grpc.Dial(host+":"+port, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{grpcChannel: grpcChannel, tables: make(map[fieldId]*TableHandle)}
+
+	client.sessionStub, err = newSessionStub(ctx, client)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.tableStub, err = newTableStub(client)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.consoleStub, err = newConsoleStub(ctx, client, scriptLanguage)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.flightStub, err = newFlightStub(client, host, port)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.inputTableStub = newInputTableStub(client)
+
+	client.appStub = newAppStub(client)
+
+	return client, nil
+}
+
+// Checks if the client is closed, i.e. it can no longer perform operations on the server.
+func (client *Client) Closed() bool {
+	return client.grpcChannel == nil
+}
+
+type FetchOption int
+
+const (
+	FetchOnce      FetchOption = iota // Fetches the list of tables once and then returns.
+	FetchRepeating FetchOption = iota // Starts up a background thread to continually update the list of tables as changes occur.
+)
+
+// Fetches the list of tables from the server.
+// This allows the client to see the list of named global tables on the server,
+// and thus allows it to open them using OpenTables.
+// Tables created in scripts run by the current client are immediately visible and do not require a FetchTables call.
+func (client *Client) FetchTables(ctx context.Context, opt FetchOption) error {
+	if client.Closed() {
+		return ErrClosedClient
+	}
+
+	err := client.listFields(ctx, func(update *applicationpb2.FieldsChangeUpdate) {
+		client.handleFieldChanges(update)
+	})
```

Comment:
Does this have good concurrency behavior?

Reply:
handleFieldChanges is thread safe.  Are there any possible race conditions, since we are calling `FetchTables` frequently?  For example, could multiple calls be made, but we do not process the events in the right order?  Is there a sequence ID or such to avoid this?

Reply:
The hypothetical race I am considering is between multiple `listFields` calls.  If call 2 executes `handleFieldChanges` before call 1, there could be an out-of-order result problem.

Reply:
I have now made `listFields` properly block on waiting for the previous call to be canceled. I believe that means it is now completely impossible for two goroutines to ever call `handleFieldChanges` at the same time.